bridge/sam/pool.go

```diff
@@ -0,0 +1,399 @@
+package sam
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+type pool struct {
+	// write-lock is called only when changing validators process
+	// otherwise read-lock is called
+	// Changing validators will occur the most rarely (once per epoch)
+	changeValidatorsLock sync.RWMutex
+	validators           []types.Address
+	threshold            uint64 // required number of signatures for ready
+
+	// storage for message body
+	messageMap sync.Map // types.Hash -> []byte
+
+	// map from message Hash -> bool for message flags
+	consumedMap sync.Map
+	readyMap    sync.Map
+
+	// signatures
+	messageSignatures *messageSignaturesStore
+}
+
+// diffAddresses returns a list of the addresses that are in arr1 but not in arr2
+func diffAddresses(arr1, arr2 []types.Address) []types.Address {
+	arr2Map := make(map[types.Address]bool)
+	for _, addr := range arr2 {
+		arr2Map[addr] = true
+	}
+
+	diff := make([]types.Address, 0)
+
+	for _, addr := range arr1 {
+		if !arr2Map[addr] {
+			diff = append(diff, addr)
+		}
+	}
+
+	return diff
+}
+
+func NewPool(validators []types.Address, threshold uint64) Pool {
+	return &pool{
+		changeValidatorsLock: sync.RWMutex{},
+		validators:           validators,
+		threshold:            threshold,
+		messageMap:           sync.Map{},
+		consumedMap:          sync.Map{},
+		readyMap:             sync.Map{},
+		messageSignatures:    newMessageSignaturesStore(),
+	}
+}
+
+// Add adds new message with the signature to pool
+func (p *pool) AddMessage(message *Message) {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	if p.hasConsumed(message.Hash) {
+		// we do no longer put the signature if the message has been consumed
+		return
+	}
+
+	p.setMessageBody(message.Hash, message.Body)
+	p.tryToPromote(message.Hash)
+}
+
+// Add adds new message with the signature to pool
+func (p *pool) AddSignature(signature *MessageSignature) {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	if p.hasConsumed(signature.Hash) {
+		// we do no longer put the signature if the message has been consumed
```

Comment:
Typo here, should be `we no longer put the signature....`

Reply:
Fixed at c92a67bf10bc73bcdf8929dca389a8e53020bf04